### ISSUE TITLE
State of Adv. Data Structures/Algos in Beta

### ISSUE THREAD

**Issue by bonham000** [freeCodeCamp/freeCodeCamp#12757]
## This Issue is for tracking improvements to the Advanced Algorithms and Data Structures Section in Beta

**[The original draft of all of these challenges can be found here.](https://github.com/FreeCodeCamp/CurriculumExpansion/issues/17#issuecomment-268102890)**

---

### Currently the following needs to be done:

**Tests:**
- [x] Review tests for Sets challenges from 'union' challenge to end and verify solutions pass.
- [ ] Review all tests for Linked List section and verify solutions pass.
- [ ] Review/Revise Circular Queue Challenge and verify solutions.
- [ ] Review tests for Graphs DFS & BFS / verify solutions.

**Formatting:**
- [ ] Review all challenges and ensure they are in the correct format, i.e. `<code>` tags, `<blockquote>` tags for code examples, etc. for the challenge content, instructions, and tests (many challenges need this).
- [ ] Review all challenges for any modifications to content.
- [ ] Standardize langauge.
- [ ] Review seed code to avoid linting errors when code is seeded.
- [ ] The GIFs in the Graph challenges may need some revision.
- [ ] The seed code for some challenges is extremely verbose.

---

### What Has Been Done:

- [x] Challenge sequence has been re-arranged and repeated challenges have been removed.
- [x] Moved all sorting algorithms to adv. algorithms section and verified that the tests work. A note on these: currently the tests are only verifying the sorting behavior, not in any way verifying that the intended sorting method was implemented correctly. Writing tests to evaluate this is certainly open for discussion here.
- [x] Solution code correctly passes the tests for all of the following: **Typed Arrays**, **Maps**, and **Trees**.

---

### Notes:

- Currently trying to progress to the next challenge after passing the last challenge in the curriculum produces erroneous behavior.
- There's still space for more advanced algorithms to be added! For instance, recursion, dynamic programming, or generally just some more difficult algorithm challenges would be interesting to add here for campers to work on.

**Comment by TheeMattOliver** [comment]
@bonham000 Hey Sean, I'm not sure if you're assigning issues, but if you are I'd like to take a crack at verifying Linked Lists solutions are passing all tests.

**Comment by bonham000** [comment]
I'm not exactly assigning but I think anyone wanting to work on these challenges should discuss it here so we don't get in each other's way! Currently, I'm been waiting on my last PR to be merged before I continue working on these.

But yes it would be great if you want to try to verify the Linked List solutions are working. I found a lot of problems with some of the other challenge tests/solutions so it's possible that there may be many issues with these as well.

**Comment by TheeMattOliver** [comment]
Cool, I'll jump on those and if I run into any trouble or power through them and can pick up another task to lighten your load I'll report back!

**Comment by bonham000** [comment]
Ok! By the way, when my PR is merged that challenge seed file is going to look totally different so it may be hard to start a contribution to it right now before it gets merged!

**Comment by bonham000** [comment]
Fyi @TheeMattOliver I just re-submitted my prior PR with a few additional changes to these sections. Once it is merged then by all means take a look at reviewing the Linked List challenges/solutions/tests if you still want to!

**Comment by TheeMattOliver** [comment]
Hey there, definitely!

I'm trying to write a new test bc the remove method doesn't actually work the way it should.

**Comment by bonham000** [comment]
@TheeMattOliver sounds great! Feel free to message me on Gitter if you want to coordinate more about getting these tests working, I know that when I was looking through the challenges I realized some of the challenges had really inadequate tests.

**Comment by TheeMattOliver** [comment]
cool, will do.

**Comment by TheeMattOliver** [comment]
Just waiting...

**Comment by bonham000** [comment]
Hey @TheeMattOliver it's been merged! Looks like we should be good to go for continued improvements here. If you want to take a look at the Linked List challenges I will confine my changes to the Sets, Circular Queue, and Graphs challenges.

**Comment by TheeMattOliver** [comment]
Great!

**Comment by erictleung** [comment]
I've slowly started checking the formatting of the text in the challenges I've written. I haven't started doing functionality testing by testing if the tests and solutions work.

**Comment by bonham000** [comment]
@erictleung nice thanks for helping review these! Your challenges look really good. 

When I initialized checked I think that only the tests for Graphs BFS and DFS didn't work, but I think for all the other challenges you submitted the provided solutions passed all the tests.

Also, the point I wrote about the GIFs for the Graph challenges was mainly about the sizing of the images, which I'm sure you'll see in Beta when you look at them if you haven't already. Might be able to fix this with inline styles?

**Comment by erictleung** [comment]
@bonham000 thanks! I'm just about to make a PR for BFS and found an error in one of the tests and fixed it. I probably missed something similar in the DFS challenge. I've added `img-responsive` to the BFS challenge but I think it is more needed in the DFS challenge. I'll see if it helps for the DFS challenge soon.

**Comment by Ethan-Arrowood** [comment]
Can I give a crack at this todo item? 
> Review tests for Sets challenges from 'union' challenge to end and verify solutions pass.

**Comment by bonham000** [comment]
@Ethan-Arrowood yes that's a great starting point. You can find the original drafts of the challenges (with solution code) in the link at the top. I think we are trying to add passing solutions to the all the tests under the `solutions: []` entry for each challenge (not all challenge have this filled in but a few do, which you can look at as an example). This is for the purposes of running tests on the code.

Also, know that some of these challenges don't have great tests or solutions written (I know by looking through them, but have not gone through and corrected all of them). So if you feel confident that some part of the challenge needs to change (i.e. intro text, instructions, or tests, etc.) feel free to change it and we can take a look at the PR.

On a few I've reviewed I've changed large parts of the challenges. Good luck!

Btw, if you're working on this feel free to message me on Gitter it'll be easier to communicate about specific questions.

**Comment by Ethan-Arrowood** [comment]
@bonham000 Great. Thank you for all the great information. I'll make sure to message you on Gitter if I run into any issues! Thanks again 👍

**Comment by Ethan-Arrowood** [comment]
@bonham000 I've been making some great changes so far. Tell me what you think of tests like [these](https://github.com/Ethan-Arrowood/freeCodeCamp/blob/set-challenges-review/seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json#L679). I plan to go back through the other challenges and add similar tests that also act as decent examples. Similarly, can you expand on the `solutions` part of each of these challenges? They run the same with or without them so I'm just unsure what their purpose is. Also what is name of Gitter I tried searching bonham000 and Sean Smith but i couldn't find you.

**Comment by bonham000** [comment]
@Ethan-Arrowood the main thing with the tests I think is that they really verify if the user has completed whatever functionality they are testing and provide a useful message about what the test is. Overall, there is not a ton of consistency with the way the tests are written in these challenges so there is always room for improvement but the minimumw goal would be just to have functioning tests that correctly verify the solution.

With the `solution` it is for running `npm test` locally to execute tests on the FCC code. So I believe if there is solution code it will try to run this against the challenge code itself and see if it passes the tests. So adding a solution code which passes the tests is a good safeguard against any future updates to the challenge seed code or tests, and I think preferably should be added anytime someone works on one of these challenges where the solution has not been added yet.

**Comment by Ethan-Arrowood** [comment]
@bonham000 you can go ahead and check off 

> Review tests for Sets challenges from 'union' challenge to end and verify solutions pass.

My commit with these challenges was just merged 😄

**Comment by bonham000** [comment]
Ok, awesome @Ethan-Arrowood! I'll check it off. I just looked at the PR, nice job!

**Comment by Ethan-Arrowood** [comment]
@bonham000 thank you 😄

**Comment by Ethan-Arrowood** [comment]
@bonham000 for linked lists should we use the list.length() method or list.size() method to get the number of elements? There are discrepancies throughout all the challenges causing a bunch of test conflicts. I just need to know what the standard should be and I'll make the change.

**Comment by QuincyLarson** [comment]
I'm closing this issue as stale since it hasn't been active lately. If you think this is still relevant to the newly updated platform, please explain why, then reopen it.
